pkg/providers/version/version.go

```diff
@@ -30,35 +36,94 @@ const (
 	kubernetesVersionCacheKey = "kubernetesVersion"
 )
 
+type HTTPClient interface {
+	Get(string) (*http.Response, error)
+}
+
 // Provider get the APIServer version. This will be initialized at start up and allows karpenter to have an understanding of the cluster version
 // for decision making. The version is cached to help reduce the amount of calls made to the API Server
-
 type Provider struct {
-	cache               *cache.Cache
-	cm                  *pretty.ChangeMonitor
-	kubernetesInterface kubernetes.Interface
+	cache      *cache.Cache
+	cm         *pretty.ChangeMonitor
+	kubeClient client.Client
+	httpClient HTTPClient
 }
 
-func NewProvider(kubernetesInterface kubernetes.Interface, cache *cache.Cache) *Provider {
+func NewProvider(cache *cache.Cache, client client.Client, httpClient HTTPClient) *Provider {
 	return &Provider{
-		cm:                  pretty.NewChangeMonitor(),
-		cache:               cache,
-		kubernetesInterface: kubernetesInterface,
+		cm:         pretty.NewChangeMonitor(),
+		cache:      cache,
+		kubeClient: client,
+		httpClient: httpClient,
 	}
 }
 
 func (p *Provider) Get(ctx context.Context) (string, error) {
 	if version, ok := p.cache.Get(kubernetesVersionCacheKey); ok {
 		return version.(string), nil
 	}
-	serverVersion, err := p.kubernetesInterface.Discovery().ServerVersion()
+	version, err := p.GetMinKubernetesVersion(ctx)
 	if err != nil {
 		return "", err
 	}
-	version := fmt.Sprintf("%s.%s", serverVersion.Major, strings.TrimSuffix(serverVersion.Minor, "+"))
 	p.cache.SetDefault(kubernetesVersionCacheKey, version)
 	if p.cm.HasChanged("kubernetes-version", version) {
 		logging.FromContext(ctx).With("version", version).Debugf("discovered kubernetes version")
 	}
 	return version, nil
 }
+
+// GetMinKubernetesVersion ensures that we query all known APIServers for the K8s version.
+// This to handle any scenarios where there may be multiple APIServer reporting different
+// K8s versions.
+func (p *Provider) GetMinKubernetesVersion(ctx context.Context) (string, error) {
+	var endpointSlice v1.EndpointSlice
+	if err := p.kubeClient.Get(ctx, types.NamespacedName{Name: "kubernetes", Namespace: "default"}, &endpointSlice); err != nil {
+		return "", fmt.Errorf("getting endpoints, %w", err)
+	}
+	// This error should never occur
+	minVersion, err := version.ParseGeneric("v99.99.99")
+	if err != nil {
+		return "", fmt.Errorf("parsing version, %w", err)
+	}
+	for _, address := range getAddresses(endpointSlice) {
+		resp, err := p.httpClient.Get(address)
```

Comment:
Does it make sense to use the default config and just update the host to be equal to the IP and port of the APIServer so that we are still using the same config and clientSet concepts rather than building the http client ourselves: https://github.com/kubernetes/client-go/blob/v0.28.1/rest/config.go#L57

Reply:
We need to use the config that's initialized as part of the Operator IIUC. 
> We could alternatively have used the simpler client-go InClusterConfig() method. However, that only works when Karpenter is running as a Pod within the same cluster it's managing.